zutil_p.h

```diff
@@ -5,30 +5,44 @@
 #ifndef ZUTIL_P_H
 #define ZUTIL_P_H
 
-#ifdef __APPLE__
-#  include <stdlib.h>
+#if !defined(MAL_IMPL) && !defined(_WIN32)
+#  if __STDC_VERSION__+0 >= 201112L
+#    define MAL_IMPL 2 /* ISO C11 `aligned_alloc` */
```

Comment:
Having the `MAL_IMPL` with number codes I think hurts readability. Perhaps we can use something more descriptive and also get away from using numbers to represent the different methods. I would be in favor of not having an intermediary `MAL_IMPL`.

Reply:
But I need a way to force a certain implementation from the build system. For Apple, I need to explicitly use the POSIX (MAL_IMPL=1) implementation. For Android, depending on the API level, I need to explicitly use either C11 (MAL_IMPL=2) or POSIX (MAL_IMPL=1) or the fallback `memalign()` (MAL_IMPL=0). If you have a cleaner way, I’m hearing.

Reply:
How about just using separate names for them? `HAVE_POSIX_MEMALIGN` or `HAVE_ALIGNED_ALLOC`, etc..?

Reply:
But then you would have something like this:
```
#if !defined(USE_ALIGNED_ALLOC) && !defined(USE_POSIX_MEMALIGN) && !defined(USE_MEMALIGN) && !defined(WIN32)
...
```
which is way too long. Also, I don’t have to set `HAVE_ALIGNED_ALLOC`, `HAVE_POSIX_MEMALIGN` and `HAVE_MEMALIGN` by using my method. I want the code to be platform (unix) agnostic (except for Windows) and only the build systems should problematic platforms like macOS and to a lesser extent Android.

Reply:
But we don’t need `HAVE_...` for standard functions (ISO C or POSIX)! If a system does not implement standard functions, then they should be handled in the build system, not the code. I don’t want my code to be dependent on the presence of a build system. That’s why I have auto-detection based on `__STDC_VERSION__` and `_POSIX_C_SOURCE`.

Reply:
You can't detect by value of ```_POSIX_C_SOURCE``` as it is defined as minimum supported version, not maximum... The build system must detect what is the maximum supported value for the compiler and then add it to compiler command line.

Reply:
@mtl1979 I already add to compiler command: https://github.com/zlib-ng/zlib-ng/pull/884/files#diff-f673bdec1661e9b36c026093309afc4ce1086fbb900f9697fb6ecaaa95c27386R30.

Reply:
@concatime I see conflicting definitions of ```_POSIX_C_SOURCE``` in configure now... In other projects we only define ```-D_GNU_SOURCE -D_DEFAULT_SOURCE``` which selects the latest supported ```_POSIX_C_SOURCE``` value on all modern compilers... The define name was recently changed, so both should be and can be safely defined.